File: rtiow/JS/main05_2.js
function hit_sphere(center, radius, r) {
    const oc = vsub(r.orig, center);
    const a = dot(r.dir, r.dir);
    const b = 2.0 * dot(oc, r.dir);
    const c = dot(oc, oc) - radius * radius;
    const discriminant = b * b - 4 * a * c;
    return discriminant > 0.0;
}
function ray_color(r) {
    if (hit_sphere(new Float32Array([0, 0, -1]), 0.5, r)) {
        return new Float32Array([1, 0, 0]);
    }
    const unit_direction = unit_vector(r.dir);
    const t = 0.5 * (unit_direction[1] + 1.0);
    return new Float32Array([1.0 - t + 0.5 * t, 1.0 - t + 0.7 * t, 1.0]);
}
function main05_2() {
    const func = arguments.callee.name || "anonymous";
    cls();
    println(`${func}`);
    var canvas = document.getElementById('canvas');
    const w = canvas.width;
    const h = canvas.height;
    if (canvas.getContext) {
        var ctx = canvas.getContext('2d', { alpha: false });
        image = ctx.getImageData(0, 0, w, h);
        const aspect_ratio = 16.0 / 9.0;
        const image_width = w;
        const image_height = h;
        const viewport_height = 2.0;
        const viewport_width = aspect_ratio * viewport_height;
        const focal_length = 1.0;
        const origin = new Float32Array([0, 0, 0]);
        const horizontal = new Float32Array([viewport_width, 0, 0]);
        const vertical = new Float32Array([0, viewport_height, 0]);
        const lower_left_corner = vsub(vsub(vsub(origin, vdiv(horizontal, 2.0)), vdiv(vertical, 2.0)), new Float32Array([0, 0, focal_length]));
        println(`origin=${origin}`)
        println(`horizontal=${horizontal}`)
        println(`vertical=${vertical}`)
        println(`lower_left_corner=${lower_left_corner}`)
        for (let j = 0; j < h; j++) {
            // println(`Scanlines remaining: ${h - j - 1}`);
            for (let i = 0; i < w; i++) {
                const u = parseFloat(i) / (w - 1);
                const v = parseFloat(h - 1 - j) / (h - 1);
                const r = { orig: origin, dir: vsub(vadd(vadd(lower_left_corner, vmul(u, horizontal)), vmul(v, vertical)), origin) };
                const pixel_color = ray_color(r);
                image.data[(j * w + i) * 4 + 0] = 255.999 * pixel_color[0];
                image.data[(j * w + i) * 4 + 1] = 255.999 * pixel_color[1];
                image.data[(j * w + i) * 4 + 2] = 255.999 * pixel_color[2];
            }
        }
        ctx.putImageData(image, 0, 0);
        println(`Done. ${func}`);
    }
}

File: rtiow/JS/C/main05_2.c
#include "conf.h"
#include <stdio.h>

#include "ray.h"
#include "vec3.h"

static int hit_sphere(const vec3 center, scalar radius, const ray r) {
  vec3 oc;
  vsub(oc, r.origin, center);
  scalar a = vdot(r.direction, r.direction);
  scalar b = 2.0 * vdot(oc, r.direction);
  scalar c = vdot(oc, oc) - radius * radius;
  scalar discriminant = b * b - 4.0 * a * c;
  return discriminant > 0.0;
}

static void ray_color(vec3 col, const ray r) {
  if (hit_sphere((vec3){0.0, 0.0, -1.0}, 0.5, r)) {
    col[0] = 1.0;
    col[1] = 0.0;
    col[2] = 0.0;
    return;
  }
  vec3 unit_direction;
  unit_vector(unit_direction, r.direction);
  scalar t = 0.5 * (unit_direction[1] + 1.0);
  vec3 col0 = {1.0, 1.0, 1.0};
  vec3 col1 = {0.5, 0.7, 1.0};
  vmul(col0, 1.0 - t, col0);
  vmul(col1, t, col1);
  vadd(col, col0, col1);
}

void main05_2(conf_t c) {
  // Image
  const int image_width = IMG_WIDTH(c);
  const int image_height = IMG_HEIGHT(c);
  const scalar aspect_ratio = (scalar)image_width / image_height;
  // Camera
  const scalar viewport_height = 2.0;
  const scalar viewport_width = aspect_ratio * viewport_height;
  const scalar focal_length = 1.0;
  // printf("viewport_width=%f\n", viewport_width);
  // printf("viewport_height=%f\n", viewport_height);
  vec3 origin = {0.0, 0.0, 0.0};
  vec3 horizontal = {viewport_width, 0.0, 0.0};
  vec3 vertical = {0.0, viewport_height, 0.0};
  vec3 lower_left_corner;
  vec3 vtmp = {0.0, 0.0, focal_length};
  vsub(lower_left_corner, origin, vtmp);
  vdiv(vtmp, horizontal, 2.0);
  vsub(lower_left_corner, lower_left_corner, vtmp);
  vdiv(vtmp, vertical, 2.0);
  vsub(lower_left_corner, lower_left_corner, vtmp);
  // Render
  uint8_t *p = PIXELS(c);
  for (int j = image_height - 1; j >= 0; --j) {
    // printf("Scanlines remaining: %d\n", j);
    for (int i = 0; i < image_width; ++i) {
      vec3 pixel_color;
      scalar u = (scalar)i / (image_width - 1);
      scalar v = (scalar)j / (image_height - 1);
      vec3 direction, direction0;
      vsub(direction, lower_left_corner, origin);
      vmul(direction0, u, horizontal);
      vadd(direction, direction, direction0);
      vmul(direction0, v, vertical);
      vadd(direction, direction, direction0);
      ray r;
      rmake(&r, origin, direction);
      ray_color(pixel_color, r);
      int ir = 255.999 * pixel_color[0];
      int ig = 255.999 * pixel_color[1];
      int ib = 255.999 * pixel_color[2];
      *p++ = ir;
      *p++ = ig;
      *p++ = ib;
    }
  }
}
